runtime/lua/vim/lsp/compat.lua

```diff
@@ -0,0 +1,12 @@
+-- Lsp forward-compatibility layer.
+--
+-- allows plugins to wrap 0.5 handler calls to use the 0.6/0.5.1 signature
+local M = {}
+
+M.wrap_05_handler = function(handler)
+  return function(err, result, context, config)
+    return handler(err, context.method, result, context.client_id, context.bufnr, config)
```

Comment:
this looks like something that can be copy-pasted into any project. So we can document it, but we don't need to introduce this compat module into the API and then later deal with removing it (thus duplicating  the migration issue we're already trying to deal with now)

Reply:
I was just thinking how to adapt the handlers in my plugins so that they work with all three: 0.5, master prior this change and master with this change.

And that would require some kind of feature or version check.
With the current PR state I'd probably create something like this:


```lua
function M.mk_handler(fn)
  if vim.lsp.compat then
    return vim.lsp.compat.wrap_05_handler(fn)
  end
  return fn
end
```

Implementing the `wrap_05_handler` isn't the issue, but some kind of marker that can be checked at runtime to see which signature format is used would be good. `nvim-0.6` won't be granular enough, as it would also return true for versions after 0.5, but before this change.

Reply:
> `nvim-0.6` won't be granular enough, as it would also return true for versions after 0.5, but before this change.

? `:echo has('nvim-0.5.1')` works

Reply:
`vim.fn.has('nvim-0.5.1')` would also return true for the current master (without this change, using the old signature) 

I guess it is acceptable that master/plugin combinations are broken for a bit, but if there were some check for patch-level or signature-format level plugins could make the transition seamless. (At least if they adapt the code before this PR is merged)

Reply:
Something like the following might work to keep compatiblity with all 3 variants:

```
function M.mk_handler(fn)
  return function(...)
    local is_new = type(select(4, ...)) == 'table'
    if is_new then
      fn(...)
    else
      local err = select(1, ...)
      local method = select(2, ...)
      local result = select(3, ...)
      local client_id = select(4, ...)
      local bufnr = select(5, ...)
      local config = select(6, ...)
      fn(err, result, { method = method, client_id = client_id, bufnr = bufnr }, config)
    end
  end
end
```